File: examples/dekita/.claude/hooks/pr-merge-pull-reminder.py
#!/usr/bin/env python3
# - 責務: PRマージ後にmainブランチを自動pull
# - 重複なし: worktree-auto-cleanupはworktree削除、本フックはmain pull
# - 自動化型: マージ成功後に即座にmainを最新化
"""
PostToolUse hook to auto-pull main after PR merge.

When `gh pr merge` succeeds, this hook:
1. Detects if the current directory is the main repository
2. Pulls the main branch to sync with remote

This prevents working with stale main branch after merging PRs.
"""

import os
import subprocess
from pathlib import Path

from lib.constants import TIMEOUT_LIGHT, TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.hook_input import get_exit_code, get_tool_result
from lib.repo import get_repo_root, is_merge_success
from lib.session import parse_hook_input


def _get_current_branch(repo_root: Path) -> str | None:
    """Get the current branch name."""
    try:
        result = subprocess.run(
            ["git", "branch", "--show-current"],
            cwd=str(repo_root),
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        if result.returncode == 0:
            return result.stdout.strip()
        return None
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return None


def is_in_worktree() -> bool:
    """Check if current directory is in a worktree."""
    cwd = os.getcwd()
    return "/.worktrees/" in cwd or cwd.endswith("/.worktrees")


def is_pr_merge_command(tool_input: str) -> bool:
    """Check if the command is a PR merge command."""
    return "gh pr merge" in tool_input


def _check_merge_success(tool_output: str, exit_code: int, command: str = "") -> bool:
    """Check if the merge was successful.

    Wrapper around common.is_merge_success for backward compatibility.
    """
    return is_merge_success(exit_code, tool_output, command)


def pull_main(repo_root: Path) -> tuple[bool, str]:
    """Pull main branch in the repository root."""
    try:
        result = subprocess.run(
            ["git", "pull", "origin", "main"],
            cwd=str(repo_root),
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode == 0:
            return True, result.stdout.strip()
        return False, result.stderr.strip()
    except subprocess.TimeoutExpired:
        return False, "Timeout while pulling main"
    except FileNotFoundError:
        return False, "git command not found"


def main() -> None:
    """Main hook logic."""
    input_data = parse_hook_input()
    if not input_data:
        return

    tool_name = input_data.get("tool_name", "")
    if tool_name != "Bash":
        return

    tool_input = input_data.get("tool_input", {})
    command = tool_input.get("command", "")

    if not is_pr_merge_command(command):
        return

    tool_output = input_data.get("tool_output", "")
    # Issue #2203: Use get_exit_code() for consistent default value
    # Use get_tool_result() to handle both tool_result and tool_response
    tool_result = get_tool_result(input_data) or {}
    exit_code = get_exit_code(tool_result)

    if not _check_merge_success(tool_output, exit_code, command):
        return

    repo_root = get_repo_root()
    if not repo_root:
        return

    # Even if in worktree, pull main in the main repository
    # Issue #727: リマインダーだけでは忘れられるため、自動実行する
    if is_in_worktree():
        # Check if main repo is on main branch before pulling
        main_repo_branch = _get_current_branch(repo_root)
        if main_repo_branch != "main":
            # Main repo is not on main branch - show reminder instead
            print(
                f"[pr-merge-pull-reminder] PRがマージされました。\n"
                f"メインリポジトリが{main_repo_branch}ブランチのため自動pullをスキップ。\n"
                f"手動でpullしてください:\n"
                f"  cd {repo_root}\n"
                f"  git checkout main\n"
                f"  git pull origin main"
            )
            log_hook_execution(
                "pr-merge-pull-reminder",
                "approve",
                f"skipped: main repo on {main_repo_branch}, not main",
            )
            return

        # Pull main in the main repo (not the worktree)
        success, output = pull_main(repo_root)
        if success:
            print(f"[pr-merge-pull-reminder] メインリポジトリでmainを自動pullしました: {output}")
            log_hook_execution(
                "pr-merge-pull-reminder",
                "approve",
                f"auto_pull from worktree: {output}",
            )
        else:
            print(f"[pr-merge-pull-reminder] main pullに失敗しました: {output}")
            log_hook_execution(
                "pr-merge-pull-reminder",
                "approve",
                f"auto_pull failed from worktree: {output}",
            )
        return

    # Auto-pull main (only if on main branch)
    current_branch = _get_current_branch(repo_root)
    if current_branch != "main":
        # Not on main branch - show reminder instead of pulling
        # (pulling origin main into a feature branch would cause issues)
        print(
            "[pr-merge-pull-reminder] PRがマージされました。"
            "mainブランチに切り替えてpullしてください: git checkout main && git pull origin main"
        )
        log_hook_execution(
            "pr-merge-pull-reminder",
            "approve",
            f"reminder: not on main (current: {current_branch})",
        )
        return

    success, output = pull_main(repo_root)
    if success:
        print(f"[pr-merge-pull-reminder] mainブランチを自動pullしました: {output}")
        log_hook_execution(
            "pr-merge-pull-reminder",
            "approve",
            f"auto_pull on main: {output}",
        )
    else:
        print(f"[pr-merge-pull-reminder] main pullに失敗しました: {output}")
        log_hook_execution(
            "pr-merge-pull-reminder",
            "approve",
            f"auto_pull failed on main: {output}",
        )


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/pr_merge_pull_reminder.ts
#!/usr/bin/env bun
/**
 * PostToolUse hook to auto-pull main after PR merge.
 *
 * Why:
 *   PRマージ後にmainブランチを最新化しないと、次のworktree作成時に
 *   古いmainをベースにしてしまう問題を防ぐ。
 *
 * What:
 *   - gh pr merge 成功を検出
 *   - mainリポジトリのmainブランチを自動pull
 *   - worktree内でも親リポジトリのmainをpull
 *
 * Remarks:
 *   - PostToolUse:Bash フック
 *   - 非ブロック型（自動pull実行、失敗時はリマインド表示）
 *   - mainブランチ以外の場合は手動pullをリマインド
 *   - Python版: pr_merge_pull_reminder.py
 *
 * Changelog:
 *   - silenvx/dekita#727: リマインダーから自動実行に変更
 *   - silenvx/dekita#2203: get_exit_code()使用に統一
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { getExitCode, getToolResult } from "../lib/input_context";
import { logHookExecution } from "../lib/logging";
import { getRepoRoot, isMergeSuccess } from "../lib/repo";
import { createContext, getSessionId, parseHookInput } from "../lib/session";

const HOOK_NAME = "pr-merge-pull-reminder";
const TIMEOUT_LIGHT = 5000;
const TIMEOUT_MEDIUM = 30000;

/**
 * Get the current branch name.
 */
export function getCurrentBranch(repoRoot: string): string | null {
  try {
    const result = execSync("git branch --show-current", {
      cwd: repoRoot,
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim() || null;
  } catch {
    return null;
  }
}

/**
 * Check if current directory is in a worktree.
 */
export function isInWorktree(): boolean {
  const cwd = process.cwd();
  return cwd.includes("/.worktrees/") || cwd.endsWith("/.worktrees");
}

/**
 * Check if the command is a PR merge command.
 * Uses regex to avoid false positives from echo/comments.
 */
export function isPrMergeCommand(command: string): boolean {
  return /^\s*gh\s+pr\s+merge\b/.test(command);
}

/**
 * Pull main branch in the repository root.
 */
function pullMain(repoRoot: string): { success: boolean; output: string } {
  try {
    const result = execSync("git pull origin main", {
      cwd: repoRoot,
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return { success: true, output: result.trim() };
  } catch (error) {
    if (error instanceof Error) {
      const execError = error as { stderr?: string; message: string };
      return {
        success: false,
        output: execError.stderr || execError.message,
      };
    }
    return { success: false, output: "Unknown error during pull" };
  }
}

async function main(): Promise<void> {
  const result: Record<string, unknown> = { continue: true };

  try {
    const data = await parseHookInput();
    const ctx = createContext(data);
    const sessionId = getSessionId(ctx) ?? "unknown";

    const toolName = data.tool_name ?? "";
    if (toolName !== "Bash") {
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) ?? {};
    const command = (toolInput.command as string) ?? "";

    if (!isPrMergeCommand(command)) {
      console.log(JSON.stringify(result));
      return;
    }

    // Get tool result for exit code and output
    const rawResult = getToolResult(data);
    const toolResult: Record<string, unknown> =
      typeof rawResult === "object" && rawResult !== null
        ? (rawResult as Record<string, unknown>)
        : {};

    const exitCode = getExitCode(rawResult);
    const stdout = String(toolResult.stdout ?? "");
    const stderr = String(toolResult.stderr ?? "");

    if (!isMergeSuccess(exitCode, stdout, command, stderr)) {
      console.log(JSON.stringify(result));
      return;
    }

    const repoRoot = getRepoRoot();
    if (!repoRoot) {
      console.log(JSON.stringify(result));
      return;
    }

    // Handle worktree case
    if (isInWorktree()) {
      const mainRepoBranch = getCurrentBranch(repoRoot);
      if (mainRepoBranch !== "main") {
        // Main repo is not on main branch - show reminder instead
        const message = `PRがマージされました。\nメインリポジトリが${mainRepoBranch}ブランチのため自動pullをスキップ。\n手動でpullしてください:\n  cd ${repoRoot}\n  git checkout main\n  git pull origin main`;

        result.systemMessage = `[${HOOK_NAME}] ${message}`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `skipped: main repo on ${mainRepoBranch}, not main`,
          undefined,
          { sessionId },
        );
        console.log(JSON.stringify(result));
        return;
      }

      // Pull main in the main repo
      const { success, output } = pullMain(repoRoot);
      if (success) {
        result.systemMessage = `[${HOOK_NAME}] メインリポジトリでmainを自動pullしました: ${output}`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `auto_pull from worktree: ${output}`,
          undefined,
          { sessionId },
        );
      } else {
        result.systemMessage = `[${HOOK_NAME}] main pullに失敗しました: ${output}`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `auto_pull failed from worktree: ${output}`,
          undefined,
          { sessionId },
        );
      }
      console.log(JSON.stringify(result));
      return;
    }

    // Not in worktree - check current branch
    const currentBranch = getCurrentBranch(repoRoot);
    if (currentBranch !== "main") {
      const message =
        "PRがマージされました。" +
        "mainブランチに切り替えてpullしてください: git checkout main && git pull origin main";
      result.systemMessage = `[${HOOK_NAME}] ${message}`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `reminder: not on main (current: ${currentBranch})`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // On main branch - auto pull
    const { success, output } = pullMain(repoRoot);
    if (success) {
      result.systemMessage = `[${HOOK_NAME}] mainブランチを自動pullしました: ${output}`;
      await logHookExecution(HOOK_NAME, "approve", `auto_pull on main: ${output}`, undefined, {
        sessionId,
      });
    } else {
      result.systemMessage = `[${HOOK_NAME}] main pullに失敗しました: ${output}`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `auto_pull failed on main: ${output}`,
        undefined,
        { sessionId },
      );
    }
  } catch {
    // Don't block Claude Code on hook failures
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
